### PR TITLE
GHC 9.0

### LINKED ISSUE #452: Guess that TypeApplications is used without being enabled
Since we don't have #56, people who have `TypeApplications` in their default-extensions get their `f @Foo` turned silently into `f@Foo`.

We can detect this by spotting `EAsPat` when rendering expressions and erroring out.

### LINKED ISSUE #480: Parse error on a standalone declaration with an associated Haddock
Hello! I'm in the process of adding `ormolu` support to HIE, but I have doubts with the way `ormolu` seems to handle formatting just a selection inside a file, or if it's even advisable!

For the moment the local tests appear to be working when formatting the whole file on save. I'm testing with [the files that `ormolu` provides as examples](https://github.com/tweag/ormolu/tree/master/data/examples).

However, when trying to format just a segment of the whole file I get some errors, but they seem a bit erratical (in some segments the formatting works, in others doesn't, in others it seems to work but then trying to format the whole file fails, as the indentation was messed up in the process).

For example, let's try with the selection made in this file (please ignore the red wiggles):

image

If I format the whole file it works without a problem. For the selection made, I think this would be the expected output:

image

When trying with Format Selection, it complains the following (emphasis mine):

> [Error - 2:59:07 PM] Request textDocument/formatting failed.
  Message: ormoluCmd: **OrmoluParsingFailed (RealSrcSpan SrcSpanOneLine "/Users/David/Desktop/ormolu-tests/examples/declaration/class/associated-data2.hs" 16 5 9) "parse error on input `data'"**
  Code: -32603 

As said, if I select other segments it works, but sometimes it will break the indentation of the files, preventing the formatting of the whole file to work correctly after that.

As I think these errors have more to do with `ormolu` than with `HIE` I have opened the issue here. Maybe this is the expected behaviour and it is discouraged to try formatting file selections?

Thanks!

### LINKED ISSUE #688: Switch to ghc-lib-parser-9.0.1.xxx
This should resolve some long-standing issues such as #474 and #480. Not sure how much work this is.

### PR DESCRIPTION
**Pull request by amesgen**
Closes #688
Closes #452
Closes #720
Closes #480

See [the comment below](https://github.com/tweag/ormolu/pull/722#issuecomment-886899094) for an overview.

### CODE REVIEW

**Comment by amesgen** [comment]
Overview:

 - There a many minor changes in the GHC AST in GHC 9.0:
    - The [new module hierarchy](https://gitlab.haskell.org/ghc/ghc/-/wikis/migration/9.0#module-hierarchy) is mostly a mechanical translation. 
    - The extension constructors now have `NoExtCon` as a strict field, and the GHC 8.8+ pattern matcher is clever enough to detect that we don't have to match it (no more `noExtCon` calls). 
    - Some unreachable constructs (e.g. `ConPatOut`) removed (less `notImplemented` calls)
    - The `RealSrcLoc` constructor of `SrcLoc` now has an extra field [`Maybe BufPos`](https://hackage.haskell.org/package/ghc-lib-parser-9.0.1.20210324/docs/GHC-Types-SrcLoc.html#t:BufPos) which indicates the raw location of a syntactic construct (so it ignores e.g. `LINE` pragmas). We simply ignore this field.
    - Due to [parsing changes in whitespace sensitivity](https://gitlab.haskell.org/ghc/ghc/-/wikis/migration/9.0#whitespace-sensitive-and-), some code that is valid in older GHC versions will no longer be able to be parsed.
      As a side effect, we can enable `TypeApplications` be default, (#452) as I am not aware of any code snippets where parsing with or without `TypeApplications` would make any difference.
    - Some language extensions got added or changed:
       - `LambdaCase` can now be used in the `proc` notation of `Arrows` [src](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/9.0.1-notes.html#arrow-notation)
       - [`QualifiedDo`](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/exts/qualified_do.html#extension-QualifiedDo) is enabled by default
       - [`LinearTypes`](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/exts/linear_types.html#extension-LinearTypes) and [`LexicalNegation`](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/exts/lexical_negation.html#extension-LexicalNegation) are disabled by default. 
 - Previously, ormolize was applied to the haskell packages of the current GHC version. Some packages have different versions at different GHC versions  (hlint, lens), so we now use the haskell packages of 8.10.4 for ormelization or all GHC versions.

Two questions to be resolved:

 1. The leksah test revealed a haddock edge case:
    ```haskell
    -- |
    --
    ---
    module Test where
    ```
    gets formatted (both with and without this PR) to
    ```haskell
    ---
    
    -- |
    module Test where
    ```
    but with this PR (due to parser changes), the AST is no longer considered to be equal (so formatting fails without `-u`).
 2. This snippet
    ```haskell
    test ::
      -- |
      a -> 
      b
    ```
    was a fixed point before this PR, but now it is formatted to
    ```haskell
    test ::
      a ->
      b
    ```
    and this works even without `-u`, as empty haddock strings no longer appear in the AST.
 
I am not sure whether these are considered to be bugs, and if so, whether they should block this PR.

**Comment by amesgen** [comment]
Update to the questions above: We decided that 1. is OK, and created #726 to track 2.

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
-  - label: Build and test with GHC 8.6.5
-    command: |
-      nix-build --keep-going --no-out-link --argstr ormoluCompiler ghc865
-    timeout: 100
   - label: Build and test with GHC 8.8.4
     command: |
       nix-build --keep-going --no-out-link --argstr ormoluCompiler ghc884
     timeout: 100
-  - label: Build and test with GHC 8.10.3
+  - label: Build and test with GHC 8.10.4
```
[the remark this answers is not in the document]
Comment (mrkkrp) [reply .buildkite/pipeline.yml:6]: Okay, I see that actually 8.10.5 is not available in nixpkgs. Let's test with 8.10.4 for now then.

File: expected-failures/leksah.txt [hunk 2]
```diff
@@ -1,3 +1,46 @@
+src/IDE/BufferMode.hs
```
[the remark this answers is not in the document]
Comment (amesgen) [reply expected-failures/leksah.txt:1]: They are due to #726, mentioned under 1. [in my comment above](https://github.com/tweag/ormolu/pull/722#issuecomment-886899094).

[the remark this answers is not in the document]
Comment (mrkkrp) [reply expected-failures/leksah.txt:1]: Ah, right. My bad!

**Review by mrkkrp — APPROVED** [review]

**Comment by Tritlo** [comment]
Question: I'm running `ghc-8.10.4` due to issues with `9.0.1` in the plugins I'm writing. Should I stay on `0.1.4.1`, or is it OK to switch to `0.2`?

**Comment by amesgen** [comment]
@Tritlo ormolu-0.2.0.0 should compile just fine on GHC 8.10.x; it is also tested [on CI](https://github.com/tweag/ormolu/blob/e0e961e39f565d4a57b312ed261c95f4e424317e/.buildkite/pipeline.yml#L6-L8).

**Comment by Tritlo** [comment]
Ok thanks! I just saw the lib-parser scroll by for 9.0.1, and was worried it might be inconsistent.
